### ISSUE TITLE
torch.log(y_pred) in NeuralNetClassifier.get_loss may cause NaN

### ISSUE THREAD

**Issue by ottonemo** [skorch-dev/skorch#481]
When I was testing the `examples/MNIST.ipynb` notebook I stumbled upon an NaN-issue whenever I changed the optimizer to RMSprop or Adam or changed the learning rate from the original value. After using the anomaly detection I saw that the underlying issue are extreme values (close to 1 or 0) in `y_pred` after softmax which resulted in `torch.log(y_pred)` yielding `-inf`. This was already a concern in one of the comments of #136 and now we have a "real-world" example of this.

There is the obvious fix of adding an epsilon value to `y_pred` before calling `torch.log` but this *feels* like justifying a bad decision. PyTorch recommends using `log_softmax` exactly for this reason and we are preventing people from using it by expecting `softmax` output. At the least we should make it easy to do it 'properly' (a flag maybe? `yes_i_want_log_softmax`, not sure about this)

Obvious fix:
```diff
diff --git a/skorch/classifier.py b/skorch/classifier.py
index 95fcad2..e955e03 100644
--- a/skorch/classifier.py
+++ b/skorch/classifier.py
@@ -52,8 +52,10 @@ class NeuralNetClassifier(NeuralNet):
             *args,
             criterion=torch.nn.NLLLoss,
             train_split=CVSplit(5, stratified=True),
+            epsilon=1e-8,
             **kwargs
     ):
+        self.epsilon = epsilon
         super(NeuralNetClassifier, self).__init__(
             module,
             *args,
@@ -102,7 +104,7 @@ class NeuralNetClassifier(NeuralNet):
     # pylint: disable=arguments-differ
     def get_loss(self, y_pred, y_true, *args, **kwargs):
         if isinstance(self.criterion_, torch.nn.NLLLoss):
-            y_pred = torch.log(y_pred)
+            y_pred = torch.log(y_pred + self.epsilon)
         return super().get_loss(y_pred, y_true, *args, **kwargs)
 
     # pylint: disable=signature-differs

```

**Comment by thomasjpfan** [comment]
Personally, I found this a blocker for using `NeuralNetClassifier`. Adding epsilon is the "quick fix", but overall I do not think we should mess with `y_pred` and pass it directly to `get_loss` and use `CrossEntropyLoss`. (This uses `log_softmax` internally)

If we can agree that this is the desired behavior, then we discuss how we can deprecate the current behavior.

**Comment by BenjaminBossan** [comment]
Don't we have those lines to fix the issue:

https://github.com/skorch-dev/skorch/blob/177a2a89c3ab414f3fa6fc1c08d61b9576ebd6e5/skorch/classifier.py#L105-L106

If you change the loss to `CrossEntropyLoss`, skorch shouldn't apply `log`.

**Comment by thomasjpfan** [comment]
Ah yes I forgot about that. What was the original motivation of doing the log automatically (when the loss is NLLoss)?

**Comment by BenjaminBossan** [comment]
The idea was that at that time, it was standard (keras, nolearn) that if you define a classifier, your output nonlinearity should be softmax.

**Comment by thomasjpfan** [comment]
@BenjaminBossan Do you think we should work toward, depreciating this behavior and try to use `log_softmax`, or use `CrossEntropyLoss`? (Either way we need to do a deprecation cycle)

**Comment by BenjaminBossan** [comment]
Not sure, maybe we should just document the behavior more prominently?

Deprecation is not free. I'm actually quite annoyed but the burst of deprecations introduced in sklearn in the recent versions, so we should only do it if really necessary.

**Comment by thomasjpfan** [comment]
I agree, deprecation is not free and should not be taken lightly. That being said, having the defaults give nans when a user changes to Adam is pretty bad. I am okay with adding the eps keyword, if we really want to avoid depreciations. Maybe not even a keyword, and add it directly and consider it a bug fix.

**Comment by githubnemo** [comment]
Unfortunately using `CrossEntropyLoss` is not an option since this would break `predict`/`predict_proba` which expects probabilities (i.e. softmax'd values).

Before discussing this issue further we should probably investigate how bad the issue is and if it can be mitigated by using smaller LRs, for example. It didn't look like it but if the problem can be circumvented by using smaller LRs without losing speed of convergence and accuracy this is also a solution to me.

**Comment by thomasjpfan** [comment]
> Unfortunately using CrossEntropyLoss is not an option since this would break predict/predict_proba which expects probabilities (i.e. softmax'd values).

Yea...This is true.

> It didn't look like it but if the problem can be circumvented by using smaller LRs without losing speed of convergence and accuracy this is also a solution to me.

With smaller learning rates, the model can still get into a state where it can be extremely confident (prob = 0 or 1) about a sample, it will just take more epochs to reach that state.

It feels like using `eps` is the only fix.

**Comment by BenjaminBossan** [comment]
> Unfortunately using `CrossEntropyLoss` is not an option since this would break `predict`/`predict_proba` which expects probabilities (i.e. softmax'd values).

This is true but fixing it should be easy for a user:

```python
class MyClassifier:
    def predict_proba(self, X):
        return np.exp(super().predict_proba(X))
```

Or maybe the user doesn't even want to have probabilities for some reason (`predict` would still work as expected). We would need to emphasize this in the docs, though.

> It feels like using `eps` is the only fix.

It certainly isn't the only fix but maybe good enough?

**Comment by thomasjpfan** [comment]
The `eps` value can be determined automatically depending on `yp`:

```py
    def get_loss(self, y_pred, y_true, *args, **kwargs):
        if isinstance(self.criterion_, torch.nn.NLLLoss):
			eps = torch.finfo(y_pred.dtype).eps
            y_pred = torch.log(y_pred + eps)
        return super().get_loss(y_pred, y_true, *args, **kwargs)
```

**Comment by BenjaminBossan** [comment]
That's too bad, I guess using clipping instead of `y_pred + eps` is also not enough?

**Comment by thomasjpfan** [comment]
Actually, it is `y_pred + eps` is enough, I must have misconfigured my kernel to point to the master version.

**Comment by Atcold** [comment]
The use of soft(arg)max not only should be throwing a warning, but the documentation should advise against using it. This is ***really*** bad practice. A frameworks should do the best thing by default, and allow the user to change that, if s/he knows what s/he is doing.

Calling a `log()` after an `exp()` is nonsense. Sorry but not sorry.
We're not doing math here. We're using floating points!

**Comment by BenjaminBossan** [comment]
I understand your argument. If you have an example where our particular implementation causes issues with numerical stability that could be avoided by using logsoftmax, please report it. Otherwise, I consider this issue to be solved in practice via the addition of eps (#491). As I mentioned, there are practical reasons not to change this.

If you want to update the documentation with a warning that this usage could theoretically lead to numerical instability and how to circumvent that, you're welcome to do that.

**Comment by Atcold** [comment]
I don't think you understand it.

`log(exp())` is not the identity function on a computer.

The warning should be issued by the framework as a mean of discouragement of following such bad practice.

**Comment by ottonemo** [comment]
Thank you for raising your concerns on this issue!

> I don't think you understand it.
> 
> `log(exp())` is not the identity function on a computer.

I think everybody in this thread is aware of this fact and nobody is trying to discuss it away. 

What Benjamin is saying is that there are *practical reasons* (`predict` interface, breaking change, deprecations, ...) and quite some discussion (see above) why we are expecting softmax instead of log-softmax in modules and that it is already possible to use, e.g. `CrossEntropyLoss` to avoid this issue altogether. The latest release also includes a change (#491) for making softmax numerically stable. We are open for suggestions and welcome any report of an actual problem (i.e., "here is my model and it breaks due to this") and possible mitigations (documentation changes, warnings, breaking changes) but please keep in mind that it is not a single dimensional problem and implementing a change is not free.

> The warning should be issued by the framework as a mean of discouragement of following such bad practice.

A warning is fine with me but there needs to be a new default and this is basically the above discussion all over again. The main issue here is that `predict_proba` expects values in `[0;1]` which would require either an additional softmax (in case of returning logits in the module) or a call to `np.exp` (in case of using log-softmax in the module). I'm open for new ideas.
